src/renderer/utils/createStorage.ts

```diff
@@ -1,73 +1,211 @@
+
+      this.onChange(newValue, oldValue);
+    });
   }
 
-  get(): T {
-    const strValue = this.storage.getItem(this.key);
+  protected onChange(value: T, oldValue?: T) {
+    if (!this.initialized) return;
 
-    if (strValue != null) {
-      try {
-        return JSON.parse(strValue);
-      } catch (e) {
-        console.error(`Parsing json failed for pair: ${this.key}=${strValue}`);
+    try {
+      if (value == null) {
+        this.storage.removeItem(this.key);
+      } else {
+        this.storage.setItem(this.key, value);
       }
-    }
 
-    return this.defaultValue;
+      this.storage.onChange?.({ value, oldValue, key: this.key });
+    } catch (error) {
+      console.error(`[change]: ${error}`, this, { value, oldValue });
+    }
   }
 
-  set(value: T) {
-    this.storage.setItem(this.key, JSON.stringify(value));
+  async load(): Promise<T> {
+    const value = await this.storage.getItem(this.key);
+    const notEmpty = value != null;
+    const notDefault = !this.isDefaultValue(value);
+
+    if (notEmpty && notDefault) {
+      this.merge(value);
+    }
 
-    return this;
+    return value;
   }
 
-  merge(value: Partial<T>) {
-    const currentValue = this.get();
+  get(): T {
+    return this.data.get();
+  }
 
-    return this.set(Object.assign(currentValue, value));
+  set(value: T) {
+    this.data.set(value);
   }
 
-  clear() {
-    this.storage.removeItem(this.key);
+  reset() {
+    this.set(this.defaultValue);
+  }
 
-    return this;
+  clear() {
+    this.data.set(null);
   }
 
-  getDefaultValue() {
-    return this.defaultValue;
+  merge(value: Partial<T> | ((draft: Draft<T>) => Partial<T> | void)) {
+    const updater = isFunction(value) ? value : (state: Draft<T>) => {
+      if (isObject(state)) return Object.assign(state, value);
+
+      return value;
+    };
+    const currentValue = this.toJS();
+    const nextValue = produce(currentValue, updater) as T;
+
+    this.set(nextValue);
   }
 
-  restoreDefaultValue() {
-    return this.set(this.defaultValue);
+  toJS() {
+    return toJS(this.get(), { recurseEverything: true });
   }
 }
+
+/**
+ * Keep intended window.localStorage state in external JSON-file.
+ * Reason: app creates random ports between restarts and as a result storage not persistent.
+ */
+export const jsonFileStorageAdapter: StorageAdapter = {
+  cwd: path.resolve((app || remote.app).getPath("userData"), "lens-local-storage"),
+  data: observable.map<string, any>([], { deep: false }),
+  initialized: false,
+
+  get filePath() {
+    // TODO: use getHostedClusterId() after moving out from cluster-store.ts
+    const clusterId = location.hostname.split(".").slice(-2, -1)[0];
+
+    return path.resolve(this.cwd, `${clusterId ?? "app"}.json`);
+  },
+
+  async init() {
+    if (this.initialized) return;
+
+    const data = await readJson(this.filePath).catch(noop);
+
+    if (data) {
+      this.data.replace(data);
+    }
+    this.bindAutoSave();
```

Comment:
this should get disposed when a cluster is removed?

Reply:
Not sure, since everything wiping out from "cluster-scope" including `iframe` (where this code actually running).